### PR TITLE
DOC: clarify Sobel transform

### LINKED ISSUE #18425: DOC: clarify that scipy.ndimage.sobel does not compute the 2D magnitude
### Describe your issue.

scipy.ndimage.sobel is produces very wrong result.
The example shown in https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.sobel.html is NOT what a Sobel filtered image looks like.
The current code reproduces the image in the example exactly (again, this is not what you want).

See https://scikit-image.org/docs/stable/api/skimage.filters.html#skimage.filters.sobel for correct implementation (though a little funny with the scaling)
Or https://en.wikipedia.org/wiki/Sobel_operator for basic implementation.

Or, use (not optimised):
    from scipy import signal
    import numpy as np
    Gx = np.array([[1, 0, -1], [2, 0, -2], [1, 0, -1]])
    Gy = Gx.T
    gradx = signal.convolve2d(img, Gx, boundary='symm', mode='same')
    grady = signal.convolve2d(img, Gy, boundary='symm', mode='same')
    G = np.sqrt(gradx**2 + grady**2)

### Reproducing Code Example

```python
from scipy import ndimage, datasets
import matplotlib.pyplot as plt
fig = plt.figure()
plt.gray()  # show the filtered result in grayscale
ax1 = fig.add_subplot(121)  # left side
ax2 = fig.add_subplot(122)  # right side
ascent = datasets.ascent()
result = ndimage.sobel(ascent)
ax1.imshow(ascent)
ax2.imshow(result)
plt.show()
```

### Error message

```shell
Runs, but produces wrong result.
```

### SciPy/NumPy/Python version and system information

```shell
1.10.0 1.23.5 sys.version_info(major=3, minor=10, micro=9, releaselevel='final', serial=0)
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/Users/rorys/anaconda3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
lapack_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/Users/rorys/anaconda3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/Users/rorys/anaconda3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
blas_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/Users/rorys/anaconda3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
Supported SIMD extensions in this NumPy install:
    baseline = SSE,SSE2,SSE3
    found = SSSE3,SSE41,POPCNT,SSE42
    not found = AVX,F16C,FMA3,AVX2,AVX512F,AVX512CD,AVX512_KNL,AVX512_SKX,AVX512_CLX,AVX512_CNL,AVX512_ICL
```

### PR DESCRIPTION
**Pull request by yannislinardos**
#### Reference issue
Closes #18425

#### What does this implement/fix?
- It adds notes to clarify the differences between the directional Sobel operators and the magnitude.
- It gives an example of how it can be used.

### CODE REVIEW

**Review by tupui — COMMENTED** [review]
Thank you for the PR @yannislinardos!

From @rkern's comment on the issue, I think we should not do that and instead explain in a notes section that there are other conventions.

In the example section, we can also add a short example (similar to the one in the issue) showing how to obtain the equivalent of scikit-image and other.

**Comment by yannislinardos** [comment]
> Thank you for the PR @yannislinardos!
> 
> From @rkern's comment on the issue, I think we should not do that and instead explain in a notes section that there are other conventions.
> 
> In the example section, we can also add a short example (similar to the one in the issue) showing how to obtain the equivalent of scikit-image and other.

Thank you for the feedback. I will make some adjustments.

File: scipy/ndimage/_filters.py [hunk 1]
```diff
@@ -433,19 +433,46 @@ def sobel(input, axis=-1, output=None, mode="reflect", cval=0.0):
...
+    >>> import numpy as np
+
+    >>> ascent = datasets.ascent().astype('int32')
+    >>> sobel_h = ndimage.sobel(ascent, 0)  # horizontal gradient
+    >>> sobel_v = ndimage.sobel(ascent, 1)  # vertical gradient
+    >>> magnitude = np.sqrt(sobel_h**2 + sobel_v**2)
+    >>> magnitude *= 255.0 / np.max(magnitude)  # normalization
+
     >>> fig = plt.figure()
     >>> plt.gray()  # show the filtered result in grayscale
```
Comment (dschmitz89) [inline scipy/ndimage/_filters.py:456]:
```suggestion
    >>> fig = plt.figure(figsize=(12, 3))
    >>> plt.gray()  # show the filtered result in grayscale
```

Besides the figure size this looks good to me. Makes it much clearer for the average user.

Comment (yannislinardos) [reply scipy/ndimage/_filters.py]: Thanks for the suggestion

File: scipy/ndimage/_filters.py [hunk 2]
```diff
@@ -433,19 +433,43 @@ def sobel(input, axis=-1, output=None, mode="reflect", cval=0.0):
...
     >>> from scipy import ndimage, datasets
     >>> import matplotlib.pyplot as plt
-    >>> fig = plt.figure()
+    >>> import numpy as np
+    >>> ascent = datasets.ascent().astype('int32')
+    >>> sobel_h = ndimage.sobel(ascent, 0)  # horizontal gradient
+    >>> sobel_v = ndimage.sobel(ascent, 1)  # vertical gradient
+    >>> magnitude = np.sqrt(sobel_h**2 + sobel_v**2)
+    >>> magnitude *= 255.0 / np.max(magnitude)  # normalization
+    >>> fig = plt.figure(figsize=(12, 3))
```
Comment (j-bowhay) [inline scipy/ndimage/_filters.py:453]:
```suggestion
    >>> fig, axs = plt.subplots(2, 2, figsize=(8, 8))
```

File: scipy/ndimage/_filters.py [hunk 3]
```diff
@@ -433,19 +433,43 @@ def sobel(input, axis=-1, output=None, mode="reflect", cval=0.0):
...
     >>> plt.gray()  # show the filtered result in grayscale
-    >>> ax1 = fig.add_subplot(121)  # left side
-    >>> ax2 = fig.add_subplot(122)  # right side
-    >>> ascent = datasets.ascent()
-    >>> result = ndimage.sobel(ascent)
+    >>> ax1 = fig.add_subplot(141)
+    >>> ax2 = fig.add_subplot(142)
+    >>> ax3 = fig.add_subplot(143)
+    >>> ax4 = fig.add_subplot(144)
     >>> ax1.imshow(ascent)
```
Comment (j-bowhay) [inline scipy/ndimage/_filters.py:459]:
```suggestion
```

File: scipy/ndimage/_filters.py [hunk 4]
```diff
@@ -433,19 +433,43 @@ def sobel(input, axis=-1, output=None, mode="reflect", cval=0.0):
...
+    >>> ax1.set_axis_off()
+    >>> ax2.imshow(sobel_h)
+    >>> ax2.set_title("horizontal")
+    >>> ax2.set_axis_off()
+    >>> ax3.imshow(sobel_v)
+    >>> ax3.set_title("vertical")
+    >>> ax3.set_axis_off()
+    >>> ax4.imshow(magnitude)
+    >>> ax4.set_title("magnitude")
+    >>> ax4.set_axis_off()
```
Comment (j-bowhay) [inline scipy/ndimage/_filters.py:470]:
```suggestion
    >>> axs[0, 0].imshow(ascent)
    >>> axs[0, 1].imshow(sobel_h)
    >>> axs[1, 0].imshow(sobel_v)
    >>> axs[1, 1].imshow(magnitude)
    >>> titles = ["original", "horizontal", "vertical", "magnitude"]
    >>> for i, ax in enumerate(axs.ravel()):
    >>>     ax.set_title(titles[i])
    >>>     ax.axis("off")
```

**Review by j-bowhay — COMMENTED** [review]
Some suggestions to streamline the plotting and make figure more readable

**Comment by yannislinardos** [comment]
> Some suggestions to streamline the plotting and make figure more readable

Thank you for the suggestions. I have committed them.

File: scipy/ndimage/_filters.py [hunk 5]
```diff
@@ -433,19 +433,35 @@ def sobel(input, axis=-1, output=None, mode="reflect", cval=0.0):
     %(mode_multiple)s
     %(cval)s
 
+    Notes
+    -----
+    This function computes the axis-specific Sobel gradient.
+    The horizontal edges can emphasised with the horizontal trasform (axis=0),
+    the vertical edges with the vertical transform (axis=1) and so on for higher dimensions.
+    These can be combined to give the magnitude.
```
Comment (j-bowhay) [inline scipy/ndimage/_filters.py:441]:
```suggestion
    the vertical edges with the vertical transform (axis=1) and so on for higher
    dimensions. These can be combined to give the magnitude.
```

**Review by j-bowhay — APPROVED** [review]
LGTM thanks @yannislinardos and congratulations on your first contribution to SciPy!

**Comment by dschmitz89** [comment]
Thanks @yannislinardos !

If anyone is still looking at this: the exaple for the [prewitt filter](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.prewitt.html#scipy.ndimage.prewitt) could use a similar update.

**Pull request merged**